activerecord/lib/arel/visitors/postgresql.rb

```diff
@@ -4,6 +4,62 @@ module Arel # :nodoc: all
   module Visitors
     class PostgreSQL < Arel::Visitors::ToSql
       private
+        def visit_Arel_Nodes_UpdateStatement(o, collector)
+          collector.retryable = false
+          o = prepare_update_statement(o)
+
+          collector << "UPDATE "
+
+          # UPDATE with JOIN is in the form of:
+          #
+          #   UPDATE t1
+          #   SET ..
+          #   FROM t2
+          #   WHERE t1.join_id = t2.join_id
+          #
+          # Or if more than one join is present:
+          #
+          #   UPDATE t1
+          #   SET ..
+          #   FROM t2
+          #   JOIN t3 ON t2.join_id = t3.join_id
+          #   WHERE t1.join_id = t2.join_id
+          if has_join_sources?(o)
```

Comment:
Not a big fan of this, it feels like `UpdateStatement` should gain a `from`, and the transformation should be done on the AST?